server/src/controllers/user.ts

```diff
@@ -3,99 +3,108 @@
 
 import { Request, Response } from 'express';
-import { Course } from './course';
+import models from '../database/models';
+import Course from '../database/models/course';
+import CourseInstance from '../database/models/courseInstance';
+import CourseTranslation from '../database/models/courseTranslation';
+import { CourseData, Language } from './course';
 
 export interface TeacherCourses {
-  current: Array<Course>,
-  previous: Array<Course>
+  current: Array<CourseData>,
+  previous: Array<CourseData>
 }
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-async function getTeacherCourses(userId: number): Promise<TeacherCourses> {
-  return {
-    current: [
-      {
-        id: 0,
-        courseCode: 'CS-E4580',
-        minCredits: 5,
-        maxCredits: 5,
-        department: {
-          fi: '',
-          sv: '',
-          en: 'Department of Computer Science'
-        },
-        name: {
-          fi: '',
-          sv: '',
-          en: 'Programming Parallel Computers D'
-        },
-        evaluationInformation: {
-          fi: '',
-          sv: '',
-          en: ''
-        }
+export async function getUserCourses(req: Request, res: Response): Promise<void> {
+  try {
+    const teacherCourses: TeacherCourses = { current: [], previous: [] };
+
+    // TODO: Go through course_role instead
+    let instances: Array<CourseInstance> = await models.CourseInstance.findAll({
+      where: {
+        responsibleTeacher: Number(req.params.userId)
       }
-    ],
-    previous: [
-      {
-        id: 1,
-        courseCode: 'ELEC-C7241',
-        minCredits: 5,
-        maxCredits: 5,
-        department: {
-          fi: '',
-          sv: '',
-          en: 'Department of Communications and Networking'
-        },
-        name: {
-          fi: 'Tietokoneverkot',
-          sv: '',
-          en: ''
-        },
-        evaluationInformation: {
-          fi: '',
-          sv: '',
-          en: ''
+    });
+
+    // Sort all courses according to end date from highest to lowest.
+    instances = instances.sort((a: CourseInstance, b: CourseInstance): number => {
+      if (a.endDate < b.endDate)
+        return 1;
+      if (a.endDate > b.endDate)
+        return -1;
+
+      return 0;
+    });
+
+    // Filter out all course instances with duplicate course IDs, leaving only
+    // the latest instance of each course, which can be used to determine
+    // whether the teacher is currently teaching the course or did so in the
+    // past.
+    const courseFlags: Set<number> = new Set();
+    instances = instances.filter((courseInstance: CourseInstance) => {
+      if (courseFlags.has(courseInstance.courseId)) {
+        return false;
+      }
+      courseFlags.add(courseInstance.courseId);
+      return true;
+    });
+
+    // Construct CourseData objects and determine whether the course is current
+    // or previous.
+    const currentDate: Date = new Date();
+    for (const i in instances) {
```

Comment:
I would avoid querying database inside for loop. Could affect performance negatively when instance count grows large.